src/ios-deploy/ios-deploy.m

```diff
@@ -1726,7 +1698,7 @@ void handle_device(AMDeviceRef device) {
 void device_callback(struct am_device_notification_callback_info *info, void *arg) {
     switch (info->msg) {
         case ADNCI_MSG_CONNECTED:
-            if(device_id != NULL || !debug || AMDeviceGetInterfaceType(info->dev) != 2) {
+            if(device_id != NULL || !debug || detect_only) {
```

Comment:
This line causes the breakage, if you revert the change it works as before (backwards compatible)

Reply:
If this is reverted of course it will be problematic since WiFi devices won't be detected (the `2` is for WiFi devices). So this if statement needs to be reworked

Reply:
Found the original PR that added that added that code block which explains the logic: https://github.com/ios-control/ios-deploy/pull/48

Reply:
Yeah, I misunderstood the if statement. It was crafted to detect wifi devices only if we don't have debug set or `device_id` is specified. 

When you specify the timeout, the `best_device_match` is used after the timeout and that is why it works. 

As the `AMDeviceGetInterfaceType(info->dev) != 2` was always true for USB devices, I think we can just remove the if and also the `best_device_match` handling. It seems that old behaviour was to use USB device and fall back to Wifi if no USB device was found. (Even if `no-wifi` is specified).

I propose rewriting the device_callback to:
```
void device_callback(struct am_device_notification_callback_info *info, void *arg) {
    switch (info->msg) {
        case ADNCI_MSG_CONNECTED:
            if (no_wifi && AMDeviceGetInterfaceType(info->dev) == 2)
            {
                NSLogVerbose(@"Skipping wifi device (type: %d)", AMDeviceGetInterfaceType(info->dev));
            }
            else
            {
                NSLogVerbose(@"Handling device type: %d", AMDeviceGetInterfaceType(info->dev));
                handle_device(info->dev);
            }
        default:
            break;
    }
}
```

Reply:
I think that the logic is not needed anymore as we use the device over wifi now as needed. It would be nice to favor USB over Wifi, but I'm not sure that we can implement that without having some delay to collect devices and then choose the best.

Reply:
I believe this is the correct logic as intended:

```c
typedef enum { DeviceInterfaceTypeUnknown, DeviceInterfaceTypeUsb, DeviceInterfaceTypeWifi } AMDeviceInterfaceType;

void device_callback(struct am_device_notification_callback_info *info, void *arg) {
    AMDeviceInterfaceType interfaceType = AMDeviceGetInterfaceType(info->dev);
    NSString* deviceTypeName = (NSString*)get_device_interface_name(info->dev);
    NSString* deviceUuid = (NSString*)AMDeviceCopyDeviceIdentifier(info->dev);

    switch (info->msg) {
        case ADNCI_MSG_CONNECTED: {
            AMDeviceConnect(info->dev);
            NSString* deviceName = (NSString*)AMDeviceCopyValue(info->dev, 0, CFSTR("DeviceName"));
            AMDeviceDisconnect(info->dev);
            NSLogVerbose(@"Device connected. '%@ (%@)' (%@)", deviceName, deviceUuid, deviceTypeName);

            // special case: find best match when device_id not specified, then break (when there is a timeout, and not detect_only). If not, we wait indefinitely
            if (_timeout > 0 && device_id == NULL && !detect_only) {
                NSLogVerbose(@"Best device match: %d", interfaceType);
                best_device_match = info->dev;
                CFRetain(best_device_match);
                break;
            }

            switch (interfaceType) {
                case DeviceInterfaceTypeUsb: {
                    NSLogVerbose(@"Handling device: '%@' (%@)", deviceName, deviceTypeName);
                    handle_device(info->dev);
                }
                break;
                case DeviceInterfaceTypeWifi: {
                    if (no_wifi || debug) { // if debug flag is on, we don't deploy for wifi connections
                        NSLogVerbose(@"Skipping device: '%@' (%@)", deviceName, deviceTypeName);
                    } else {
                        NSLogVerbose(@"Handling device: '%@' (%@)", deviceName, deviceTypeName);
                        handle_device(info->dev);
                    }
                }
                break;
                case DeviceInterfaceTypeUnknown:
                break;
            }
        }
        break;
        case ADNCI_MSG_DISCONNECTED: {
            NSLogVerbose(@"Device disconnected. '(%@)' (%@)", deviceUuid, deviceTypeName);
        }
        break;
        case ADNCI_MSG_UNKNOWN:
        break;
    }
}
```

Reply:
(my latest reply came in before I read your reply (Github didn't refresh any new comments), so there may be duplication in what you have already said)

Reply:
I may be misunderstanding, but it seems that in your variant the `if(_timeout ..)` part would cause the last device detected to be chosen and only after the timeout has expired. As now we dont have special handling for Wifi Here, I think that the best_device_match does not have any usable logic left in it currently, but the timeout part breaks actual timeout usage. 

Also, why we dont wanto deploy over wifi if debug is present? It's my understanding that this PR fixes the reasoning why it was there.

Reply:
Regarding the no debug over wifi, I was relying on the information in PR #48, I don't have experience with wifi deploys. My guess is it was put there because we can't link `lldb` over wifi -- if we can, then we can remove the restriction.

The `if(_timeout...)` code was to preserve the existing logic with the timeout and how that works (save the last detected device, for use when the timeout expires, so it can handle it). Ok, I also now see a memory leak there in the CFRetain if more than 1 device is connected before the timeout... (should release the previous if it is set)

Personally I would love to refactor this, but I am settling for "not breaking existing functionality" for now.

Reply:
Yes, running and debugging over wifi now works. 

I think you are misunderstanding the old functionallity. The `AMDeviceGetInterfaceType(info->dev) != 2` is always true for USB devices, so `best_device_match` is  set only if we are trying to debug with only Wifi device present. (it will search for USB devices, if finds none - falls back to wifi after timeout). 

Now, as the wifi debugging works,  we dont need to fallback, we can just use the first device we get.

Atleast for the Iphone I have, the USB connection is always detected first, which would be preferrable, but this may not be the case always.  For cases when you want to force the USB connection, one can still use the `--no-wifi` flag.

Reply:
Alright. Can you make the changes needed as you proposed and I'll retest.

Reply:
Hi, pushed the changes.